<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0321ENSkillsNetwork928-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="400" alt="Skills Network Logo"  />
    </a>
</p>


# **Data Wrangling Lab**


Estimated time needed: **45 to 60** minutes


In this assignment you will be performing data wrangling.


## Objectives


In this lab you will perform the following:


-   Identify duplicate values in the dataset.

-   Remove duplicate values from the dataset.

-   Identify missing values in the dataset.

-   Impute the missing values in the dataset.

-   Normalize data in the dataset.


<hr>


## Hands on Lab


Import pandas module.


In [1]:
import pandas as pd

Load the dataset into a dataframe.


<h2>Read Data</h2>
<p>
We utilize the <code>pandas.read_csv()</code> function for reading CSV files. However, in this version of the lab, which operates on JupyterLite, the dataset needs to be downloaded to the interface using the provided code below.
</p>


The functions below will download the dataset into your browser:


In [2]:
from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

In [3]:
file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv"

To obtain the dataset, utilize the download() function as defined above:  


In [49]:
await download(file_path, "m1_survey_data.csv")
file_name="m1_survey_data.csv"

Utilize the Pandas method read_csv() to load the data into a dataframe.


In [86]:
df = pd.read_csv(file_name)

> Note: This version of the lab is working on JupyterLite, which requires the dataset to be downloaded to the interface.While working on the downloaded version of this notebook on their local machines(Jupyter Anaconda), the learners can simply **skip the steps above,** and simply use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [2]:
df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv")

## Finding duplicates


In this section you will identify duplicate values in the dataset.


 Find how many duplicate rows exist in the dataframe.


In [3]:
df.duplicated().sum()

154

In [4]:
df["Respondent"].duplicated().sum()

154

## Removing duplicates


Remove the duplicate rows from the dataframe.


In [5]:
df.drop_duplicates(inplace = True)

Verify if duplicates were actually dropped.


In [6]:
df.shape

(11398, 85)

In [7]:
df["Respondent"].shape

(11398,)

In [8]:
df.duplicated().sum()

0

## Finding Missing values


Find the missing values for all columns.


In [9]:
df.isnull().sum()

Respondent        0
MainBranch        0
Hobbyist          0
OpenSourcer       0
OpenSource       81
               ... 
Sexuality       542
Ethnicity       675
Dependents      140
SurveyLength     19
SurveyEase       14
Length: 85, dtype: int64

Find out how many rows are missing in the column 'WorkLoc'


In [10]:
df["WorkLoc"].isnull().sum()

32

In [11]:
df['EdLevel'].isnull().sum()

112

In [12]:
df['Country'].isnull().sum()

0

## Imputing missing values


Find the  value counts for the column WorkLoc.


In [13]:
df["WorkLoc"].value_counts()

WorkLoc
Office                                            6806
Home                                              3589
Other place, such as a coworking space or cafe     971
Name: count, dtype: int64

Identify the value that is most frequent (majority) in the WorkLoc column.


In [14]:
# Office

Impute (replace) all the empty rows in the column WorkLoc with the value that you have identified as majority.


In [15]:
df["WorkLoc"].fillna(value = "Office", inplace = True)

/tmp/ipykernel_53898/266539921.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["WorkLoc"].fillna(value = "Office", inplace = True)


After imputation there should ideally not be any empty rows in the WorkLoc column.


Verify if imputing was successful.


In [16]:
df["WorkLoc"].isnull().sum()

0

In [17]:
df["Employment"].value_counts()

Employment
Employed full-time    10968
Employed part-time      430
Name: count, dtype: int64

In [18]:
df["UndergradMajor"].value_counts()

UndergradMajor
Computer science, computer engineering, or software engineering          6953
Information systems, information technology, or system administration     794
Another engineering discipline (ex. civil, electrical, mechanical)        759
Web development or web design                                             410
A natural science (ex. biology, chemistry, physics)                       403
Mathematics or statistics                                                 372
A business discipline (ex. accounting, finance, marketing)                244
A social science (ex. anthropology, psychology, political science)        210
A humanities discipline (ex. literature, history, philosophy)             207
Fine arts or performing arts (ex. graphic design, music, studio art)      161
I never declared a major                                                  124
A health science (ex. nursing, pharmacy, radiology)                        24
Name: count, dtype: int64

In [19]:
df.describe()

Respondent     CompTotal  ConvertedComp   WorkWeekHrs   CodeRevHrs  \
count  11398.000000  1.058900e+04   1.058200e+04  11276.000000  8972.000000   
mean   12490.392437  7.570477e+05   1.315967e+05     42.064606     4.781071   
std     7235.461999  9.705598e+06   2.947865e+05     24.672741     4.567060   
min        4.000000  0.000000e+00   0.000000e+00      3.000000     0.000000   
25%     6264.250000  2.500000e+04   2.686800e+04     40.000000     2.000000   
50%    12484.000000  6.500000e+04   5.774500e+04     40.000000     4.000000   
75%    18784.750000  1.200000e+05   1.000000e+05     43.000000     5.000000   
max    25142.000000  7.000000e+08   2.000000e+06   1012.000000    99.000000   

                Age  
count  11111.000000  
mean      30.778895  
std        7.393686  
min       16.000000  
25%       25.000000  
50%       29.000000  
75%       35.000000  
max       99.000000

## Normalizing data


There are two columns in the dataset that talk about compensation.

One is "CompFreq". This column shows how often a developer is paid (Yearly, Monthly, Weekly).

The other is "CompTotal". This column talks about how much the developer is paid per Year, Month, or Week depending upon his/her "CompFreq". 

This makes it difficult to compare the total compensation of the developers.

In this section you will create a new column called 'NormalizedAnnualCompensation' which contains the 'Annual Compensation' irrespective of the 'CompFreq'.

Once this column is ready, it makes comparison of salaries easy.


<hr>


List out the various categories in the column 'CompFreq'


In [20]:
print(df["CompFreq"].unique())

['Yearly' 'Monthly' 'Weekly' nan]


Create a new column named 'NormalizedAnnualCompensation'. Use the hint given below if needed.


Double click to see the **Hint**.

<!--

Use the below logic to arrive at the values for the column NormalizedAnnualCompensation.

If the CompFreq is Yearly then use the exising value in CompTotal
If the CompFreq is Monthly then multiply the value in CompTotal with 12 (months in an year)
If the CompFreq is Weekly then multiply the value in CompTotal with 52 (weeks in an year)

-->


In [21]:
pd.set_option('future.no_silent_downcasting', True)

df["CompFreq"].replace(to_replace="Yearly", value = 1, inplace = True)
df["CompFreq"].replace(to_replace="Monthly",value = 12, inplace = True)
df["CompFreq"].replace(to_replace="Weekly", value = 52, inplace = True)

/tmp/ipykernel_53898/3047022027.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["CompFreq"].replace(to_replace="Yearly", value = 1, inplace = True)
/tmp/ipykernel_53898/3047022027.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(val

In [22]:
df["CompFreq"].value_counts()

CompFreq
1     6073
12    4788
52     331
Name: count, dtype: int64

In [23]:
df["NormalizedAnnualCompensation"] = df["CompTotal"] * df["CompFreq"]

In [24]:
df["NormalizedAnnualCompensation"]

0         61000.0
1        138000.0
2         90000.0
3        348000.0
4         90000.0
           ...   
11547    130000.0
11548     74400.0
11549    105000.0
11550     80000.0
11551         NaN
Name: NormalizedAnnualCompensation, Length: 11398, dtype: object

In [25]:
df['NormalizedAnnualCompensation'].median()

100000.0

In [26]:
df

Respondent                      MainBranch Hobbyist  \
0               4  I am a developer by profession       No   
1               9  I am a developer by profession      Yes   
2              13  I am a developer by profession      Yes   
3              16  I am a developer by profession      Yes   
4              17  I am a developer by profession      Yes   
...           ...                             ...      ...   
11547       25136  I am a developer by profession      Yes   
11548       25137  I am a developer by profession      Yes   
11549       25138  I am a developer by profession      Yes   
11550       25141  I am a developer by profession      Yes   
11551       25142  I am a developer by profession      Yes   

                                             OpenSourcer  \
0                                                  Never   
1                             Once a month or more often   
2      Less than once a month but more than once per ...   
3                                                  Never   
4      Less than once a month but more than once per ...   
...                                                  ...   
11547                                              Never   
11548                                              Never   
11549                            Less than once per year   
11550  Less than once a month but more than once per ...   
11551  Less than once a month but more than once per ...   

                                              OpenSource          Employment  \
0      The quality of OSS and closed source software ...  Employed full-time   
1      The quality of OSS and closed source software ...  Employed full-time   
2      OSS is, on average, of HIGHER quality than pro...  Employed full-time   
3      The quality of OSS and closed source software ...  Employed full-time   
4      The quality of OSS and closed source software ...  Employed full-time   
...                                                  ...                 ...   
11547  OSS is, on average, of HIGHER quality than pro...  Employed full-time   
11548  The quality of OSS and closed source software ...  Employed full-time   
11549  The quality of OSS and closed source software ...  Employed full-time   
11550  OSS is, on average, of LOWER quality than prop...  Employed full-time   
11551  OSS is, on average, of HIGHER quality than pro...  Employed full-time   

              Country Student  \
0       United States      No   
1         New Zealand      No   
2       United States      No   
3      United Kingdom      No   
4           Australia      No   
...               ...     ...   
11547   United States      No   
11548          Poland      No   
11549   United States      No   
11550     Switzerland      No   
11551  United Kingdom      No   

                                                 EdLevel  \
0               Bachelor’s degree (BA, BS, B.Eng., etc.)   
1      Some college/university study without earning ...   
2            Master’s degree (MA, MS, M.Eng., MBA, etc.)   
3            Master’s degree (MA, MS, M.Eng., MBA, etc.)   
4               Bachelor’s degree (BA, BS, B.Eng., etc.)   
...                                                  ...   
11547        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
11548        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
11549        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
11550  Secondary school (e.g. American high school, G...   
11551          Other doctoral degree (Ph.D, Ed.D., etc.)   

                                          UndergradMajor  ...  \
0      Computer science, computer engineering, or sof...  ...   
1      Computer science, computer engineering, or sof...  ...   
2      Computer science, computer engineering, or sof...  ...   
3                                                    NaN  ...   
4      Computer science, computer engineering, or sof...  ...   
...                                                  ...  ...   
11547  

## Authors


Ramesh Sannareddy


### Other Contributors


Rav Ahuja


 ## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2024-09-24|1.1|Madhusudhan Moole|Updated lab|
|2024-09-23|1.0|Raghul Ramesh|Created lab|


<!--| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                 |
| ----------------- | ------- | ----------------- | ---------------------------------- |
| 2020-10-17        | 0.1     | Ramesh Sannareddy | Created initial version of the lab |--!>


## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>
